sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsWithStateSuite.scala

```diff
@@ -873,6 +875,96 @@ class FlatMapGroupsWithStateSuite extends StateStoreMetricsTest with BeforeAndAf
     assert(e.getMessage === "The output mode of function should be append or update")
   }
 
+  test("SPARK-21977: coalesce(1) should still be repartitioned when it has keyExpressions") {
```

Comment:
what changed code paths this test cover that is not already covered by the other tests you added.